### ISSUE TITLE
Remove implicit tearoff of `call` methods

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#2399]

**Comment by leafpetersen** [comment]
@natebosch will investigate the feasibility of this.

**Comment by kevmoo** [comment]
IIRC this would be a potential (big?) win for dart2js (@sigmundch @rakudrama ?)

**Comment by lrhn** [comment]
I don't think it will do much for dart2js since it just makes an implicit inferred `.call` into an explicit `.call`. When dart2js sees the code, inference should already have happened in the front end, so there won't be any difference.

The front-end inference might get slightly faster.

(I'd be happy to be wrong, though!)

**Comment by sigmundch** [comment]
Does this issue relate to `supportsExplicitGetterCalls` (https://github.com/dart-lang/sdk/blob/94d797ebee87e7bd02521c465e52056fdc9b4d01/pkg/kernel/lib/target/targets.dart#L503)?

Dart2js currently can't use the CFE lowering to encode getter calls as an explicit `.call` because of JS interop. For example, in code like this:
```dart
class A {
  Function get f;
}

main() => A().f();
```

The CFE lowering changes main to be equivalent to `(A().f)()`. This is enabled for the VM, but we have this disabled in dart2js. For historical reasons, the code above is implemented as a direct call, and changing it to load the function and invoke it separately unfortunately changes semantics: it drops the receiver if the property was referring to a JavaScript function.

**Comment by natebosch** [comment]
> Does this issue relate to `supportsExplicitGetterCalls`

No. This issue only relates to how non-Function types are implicitly coerced to a Function type by tearing off `.call`. There is no tearoff or non-Function type involved in the getter call scenario.

**Comment by natebosch** [comment]
I have an implementation of a fix for this usage which unconditionally wraps the expression in parenthesis to make it safe to tear off `.call`. https://dart-review.googlesource.com/c/sdk/+/256163

@bwilkerson - I am trying to figure out if there might be a way to avoid introducing these parens when they are not necessary. Do you have any suggestions for reliable ways to detect this? Would it make sense to try to enumerate the types of expressions that do or don't need wrapping?

**Comment by eernstg** [comment]
https://dart-lang.github.io/linter/lints/unnecessary_parenthesis.html.

**Comment by mit-mit** [comment]
@natebosch any updates about this?

**Comment by natebosch** [comment]
The lint is working. I need to take a look at the unnecessary_parenthesis implementation to see if it is simple enough to reproduce in the fix.

**Comment by btrautmann** [comment]
I'm having a bit of trouble understanding the intent of this ticket, could somebody help me understand if the following would still be possible:

```dart
class MyClass {
  void call({required String name}) {
    ...
  }
}

void main() {
  final myClass = MyClass();
  myClass(name: 'Brandon');
}
```
I'm hoping so, since this is one of my favorite language features coming from Kotlin.

**Comment by lrhn** [comment]
Calling the callable object's `call` method implicitly will still be possible.
What would not be possible is:
```dart
  void Function({required String name}) f = MyClass();
```
You would have to do:
```dart
  void Function({required String name}) f = MyClass().call;
```
You won't get an *implicit* *tear-off* of the `call` method in a function-typed context.

In Dart 1, the `MyClass()` object was directly assignable to the function type. Users may still expect that `f` will contain the `MyClass` instance. It won't, and that's concerning from a readability stand-point.

It's actually the context-dependency that worries me, more than the tear-off readability. If you care about the readability, you can always write `.call` yourself.
In a context like `FutureOr<Function> f = o;`, where `o` has a type which both implements `Future<Function>` *and* has a `call` method, we should still do "something reasonable". It's not always that the user actually agree with the compiler what the reasonable thing is. Currently we do *something* ... I'm not sure I even remember exactly what.

Because it sometimes get tricky, especially around union types like `FutureOr` and `X?`, which, removing the responsibility for figuring out the right thing from the compiler, and putting it on the author, seems like a worth-while change. Even though it does make authors write more code, and makes "callable objects" act less like functions.

**Comment by btrautmann** [comment]
That was extremely helpful @lrhn, thank you for that explanation! 

> Even though it does make authors write more code, and makes "callable objects" act less like functions.

Totally agree that predictability/reasonableness in this case is more valuable than readability/minimizing code written.

Thanks again!

**Comment by eernstg** [comment]
Note that the ground is being prepared for this change with a lint: [avoid-implicit-call-tearoffs](https://github.com/dart-lang/linter/pull/3592).

**Comment by natebosch** [comment]
We will want to include the lint in our core lint rules so that the widest audience sees it surfaced.

**Comment by mit-mit** [comment]
Hi @natebosch any news about the `fix` for this?

**Comment by natebosch** [comment]
> any news about the `fix` for this?

Fix is implemented and waiting for review. https://dart-review.git.corp.google.com/c/sdk/+/256163

The crbot failures in that CL are related to the `package:linter` roll and should be resolved by https://dart-review.git.corp.google.com/c/sdk/+/264961

**Comment by spkersten** [comment]
We have classes like:
```
class Function1Mock<T0, T1> extends Mock {
  T0 call([T1? arg1]) => ...
}
```
that are used in tests like:
```
late Function2Mock<void, String> onNameChanged;

setUp(() {
  onNameChanged = Function2Mock();
});

// code being tested
// void sut(Function<void, String> onNameChanged) => ...

test("...", () {
  sut(onNameChanged);
  ...
  verify(onNameChanged("expected name")).called(1);
});
```
Would `sut(onNameChanged);` need to be changed to `sut(onNameChanged.call);`?

**Comment by lrhn** [comment]
@spkersten Yes. That's exactly the situation where the change applies, an implicit tear-off of the `call` method due to a function-typed context type.

**Comment by srawlins** [comment]
It's so great that a lot was done in 2.19 for this change, with a new lint rule and a fix. Is this change planned for Dart 3.0?

**Comment by natebosch** [comment]
> Is this change planned for Dart 3.0?

No, we are going to put the lint in our core set now, and then put this change behind a language version.

There are still some open questions about the UX after this change. Once the language doesn't have implicit tearoffs, we will either need to

- Re-implement the lint to find places where a tear-off _would_ have been possible, instead of where they are happening. @bwilkerson - do you have any impression on how hard this might be? The current implementation is very simple because the AST has a node for the implicit tear-off. If the change is language versioned - would we keep that node in the AST?
- Attach a fix to the type error that will surface. @bwilkerson - is it feasible to implement a quick fix that only works for a small subset of a given diagnostic's occurrences?
- Don't offer specific help to the user. @mit-mit - how critical is it that we have quick fixes to add the `.call` tearoff which work in the language version without implicit tearoffs?

**Comment by itsjustkevin** [comment]
I am going to remove this breaking change from the Dart 3 project as it is not planned for Dart 3.

**Comment by escamoteur** [comment]
I know I'm late for this discussion but for my package flutter_command this change would make all code look much uglier as the beauty of a callable class is that it can be assigned to event handlers which is a central use case for my command package.

**Comment by rrousselGit** [comment]
I have yet to find a case where the associated lint was helpful to me. On the flip side, having to add an extra `.call` everywhere I voluntarily used a tear-off was a degradation of the syntax IMO

I don't care that `Function foo = Callbable()` doesn't enable `foo is Callbable`.
But I do use callable classes quite often for spying on event-handlers:

```dart
class OnChangeListener<T> extends Mock {
  void cal(T value);
}

test('example', () {
  final listener = OnChangeListener<int>();

  someObservableObject.listen(listener); 

  verify(listener(0));
});
```

**Comment by rrousselGit** [comment]
Food for thought: Couldn't we instead remove the implicit tear-off and rather have callable classes implement `Function`?


So we could still do `Function foo = Callable();`.
But we could also do `(foo as Callable).field`

I assume that would remove the original concern?

**Comment by escamoteur** [comment]
It just looks so ugly:
![grafik](https://github.com/dart-lang/language/assets/16635729/f5a33e51-152e-4d10-90fb-cdc0220ed57a)

**Comment by lrhn** [comment]
@rrousselGit As it is, every object which implements `Function` *is* a function value, which are a separate kind of objects from class instances. Which is not a *big* optimization, because the only thing you can do with `Function`-typed values is to do dynamic invocations on them. You can get the same effect by casting to `dynamic` before doing the dynamic invocation.

But that also means that `Function` is not very useful. (You can just use `dynamic`.)
Most likely, just implementing `Function` is not enough, and people will want the object to implement the *actual* function type of the `call` method. Which is just not compatible with Dart's current type system, when generics are covariant, but `call` method arguments would be contravariant. There's a reason we stopped doing that when moving to the sounder Dart 2 type system.

So, making callable classes implement `Function` is likely too little to be useful, yet too invasive for the type system to be  practical.

For @escamoteur, if you call your function `run` instead of `call`, then:
```dart
      onRetry: feedSource.updateDataCommand.run
```
doesn't look as bad. It looks weird because `call` isn't a good name for that function, it's only there to not be visible when calling.

**Comment by escamoteur** [comment]
It might look better but I don't know why this change is necessary at all. I never heard about anyone having problems with the way it is now and the idea that a command object can be used like a function just make sense as you can call it directly too.
I really hope you don't plan to abandon callable classes as a whole. My other package get_it lives from them too.
Am 9. Jan. 2024, 18:41 +0100 schrieb Lasse R.H. Nielsen ***@***.***>:

**Comment by escamoteur** [comment]
@mraleph this still is an questionable change and I'm currently pondering about really renaming my execution function to `run` but IMHO callable classes are really a beautiful feature but then they should be treated as normal function calls.

**Comment by mraleph** [comment]
> this still is an questionable change

As I have mentioned on Twitter this change did not actually happen. It has not been spec'ed, fully decided upon and implemented. There were no changes in this area for a while now. AFAIK the only thing that shipped is a lint warning about implicit call tear-off.

@leafpetersen have outlined the reasons for the change in the original message. I do agree with them: the current behavior is rather _unobvious_ and is a compromise we arrived to when moving from Dart 1 to Dart 2 type system.

The biggest gotcha in the current behavior is the following:

```dart
class Callable {
  void call() { }
}

Callable c = Callable();
print(c is Function); // false
Function f = c;  // ok, surprisingly despite is returning false
print(f is Callable);  // false
print(identical(f, c));  // false
```

**Comment by escamoteur** [comment]
But despite that surprising behavior that you listed, how many of them will developers really face in normal usage.
The added lint on the other site happens when you use a callable class in widget callback like any other function. It works, it does not make problems, why confusing the user with a warning?
Am 10. Nov. 2025, 06:53 -0500 schrieb Slava Egorov ***@***.***>:
